### ISSUE TITLE
08-k-nearest-neighbours-winequality-red.ipynb

### ISSUE THREAD

**Issue by mezcolantriz** [4GeeksAcademy/Eder_Zelaya-Machine-Learnig#8]
@EderZelaya 

## LO QUE HAS HECHO BIEN

- **Limpieza de datos**: Verificas duplicados, infinitos y nulos. Correcto.
- **Creación de variable label**: Transformas quality (3-8) a 3 clases (0, 1, 2). ¡Bien!
- **Train/test split con stratify**: `stratify=y` para mantener proporción de clases. Excelente.
- **Escalado CORRECTO**: `fit_transform` en train, `transform` en test. ¡Perfecto!
- **RandomizedSearchCV**: Optimizas múltiples hiperparámetros (n_neighbors, weights, p, algorithm, leaf_size).
- **best_estimator_**: Extraes el modelo optimizado correctamente.
- **Guardas el modelo Y el scaler**: ¡Muy bien! Muchos olvidan el scaler.
- **Comparación escalado vs sin escalar**: Interesante mostrar la diferencia.

---

### 1. Código desordenado y duplicado
Tienes MUCHO código repetido:

```python
# drop_duplicates 3 veces:
df.drop_duplicates(inplace=True)
df = df.drop_duplicates()
df = df.drop_duplicates()

# quality_to_label aplicado 2 veces:
df['label'] = df['quality'].apply(quality_to_label)
df['label'] = df['quality'].apply(quality_to_label)

# Dos splits diferentes:
# Primero con label, luego con quality original
```

**Esto confunde y hace difícil saber qué se está usando realmente.**

---

### 2. Haces DOS modelos diferentes sin explicar por qué
```python
# Primer modelo: con 'label' (3 clases: 0, 1, 2)
X = df.drop(['quality', 'label'], axis=1)
y = df['label']

# Segundo modelo: con 'quality' original (6 clases: 3-8)
X = df.drop("quality", axis=1)  # ¡Pero 'label' sigue en X!
y = df["quality"]
```

El segundo split tiene un **ERROR**: `label` queda incluido en X, lo cual es data leakage porque `label` se deriva de `quality`.

---

### 3. ERROR: El segundo split incluye 'label' en las features
```python
X = df.drop("quality", axis=1)  # 'label' sigue dentro de X!
```

Debería ser:
```python
X = df.drop(["quality", "label"], axis=1)
```

---

### 4. Falta el gráfico de accuracy vs k
El enunciado pedía:
> "Crea un bucle para probar diferentes valores de k. Grafica accuracy vs k"

**NO hay ningún gráfico.** Deberías añadir:

```python
k_values = range(1, 31)
accuracies = []

for k in k_values:
    knn = KNeighborsClassifier(n_neighbors=k)
    knn.fit(X_train_scaled, y_train)
    y_pred = knn.predict(X_test_scaled)
    accuracies.append(accuracy_score(y_test, y_pred))

plt.figure(figsize=(10, 6))
plt.plot(k_values, accuracies, marker='o')
plt.xlabel('Número de vecinos (k)')
plt.ylabel('Accuracy')
plt.title('Accuracy vs K en KNN')
plt.grid(True)
plt.show()
```

---

### 5. Falta confusion_matrix y classification_report
Solo calculas accuracy. Deberías añadir:

```python
from sklearn.metrics import confusion_matrix, classification_report
import seaborn as sns

# Matriz de confusión
cm = confusion_matrix(y_test, y_pred)
plt.figure(figsize=(6, 4))
sns.heatmap(cm, annot=True, fmt='d', cmap='Blues')
plt.xlabel('Predicción')
plt.ylabel('Real')
plt.title('Matriz de Confusión')
plt.show()

# Classification report
print(classification_report(y_test, y_pred))
```

---

### 6. Falta decidir qué modelo usar
Tienes dos enfoques (3 clases vs 6 clases) pero no concluyes cuál es mejor ni por qué.

---

### 7. Faltan conclusiones
No hay análisis final:
- ¿Qué k es óptimo?
- ¿Mejora el modelo con optimización?
- ¿Hay overfitting?

---
Tienes los elementos principales pero está **desordenado** con código duplicado y dos enfoques mezclados sin explicación ni el porqué tomas un camino u otro!  El error de incluir `label` en X del segundo modelo es grave. Faltan las conclusiones. Limpia el código, elige un enfoque, y añade el gráfico. ¡Y observaciones, es lo que te ayudará a saber qué estás haciendo y por qué!
